lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -891,6 +891,93 @@ OpFoldResult OrOp::fold(ArrayRef<Attribute> constants) {
       constants, [](APInt &a, const APInt &b) { a |= b; });
 }
 
+/// Merge concats in an or op when they have non-conflicting fields.
+static bool canonicalizeOrExclusiveConcat(OrOp op, PatternRewriter &rewriter) {
+  auto inputs = op.inputs();
+  auto size = inputs.size();
+  auto concat1 = inputs.back().getDefiningOp<ConcatOp>();
+  auto concat2 = inputs[size - 2].getDefiningOp<ConcatOp>();
+
+  if (!concat1 || !concat2)
+    return false;
+
+  auto knownBits1 = computeKnownBits(inputs.back());
+  auto knownBits2 = computeKnownBits(inputs[size - 2]);
+  auto unknownBits1 = ~(knownBits1.Zero | knownBits1.One);
+  auto unknownBits2 = ~(knownBits2.Zero | knownBits2.One);
+  bool anyUnknownBitsOverlap = !(unknownBits1 & unknownBits2).isNullValue();
+  bool anyConstantOne = !(knownBits1.One | knownBits2.One).isNullValue();
+
+  // We can simplify only if unknown bits do not overlap. We don't handle the
+  // case where a constant 1 is present anywhere yet.
+  if (anyUnknownBitsOverlap || anyConstantOne)
+    return false;
+
+  SmallVector<Value> newConcatOperands;
+
+  // Simultaneously iterate over the operands of both concat ops, from MSB to
+  // LSB, pushing ranges of bits out at a time. Note that when we only use up
+  // part of a constant value, we mutate the operands array with a truncated
+  // version of the constant.
+  SmallVector<Value> operands1 = concat1->getOperands();
+  SmallVector<Value> operands2 = concat2->getOperands();
+  auto *it1 = operands1.begin();
+  auto *it2 = operands2.begin();
+  auto *end1 = operands1.end();
+  auto *end2 = operands2.end();
+  while (it1 != end1 && it2 != end2) {
+    auto operand1 = (*it1);
```

Comment:
you could just getDefiningOp and not store operand[1,2].  the constantop's type would work below, as would adding it to the new operands.

Reply:
I actually end up using `operand[1,2]` in the case where one of the current operands is a constant 0 and the other is some other value, since I directly push the other, non-ConstantOp value to `newConcatOperands`.

I think what I'd really love here is something like pattern matching in a functional programming language where I extract out just the constant value in the case where it's a constant and where I only have the opaque type in the cases where it's not a ConstantOp.